framework/src/org/checkerframework/common/value/ValueVisitor.java

```diff
@@ -171,6 +173,40 @@ public Void visitTypeCast(TypeCastTree node, Void p) {
         if (node.getExpression().getKind() == Kind.NULL_LITERAL) {
             return null;
         }
+
+        AnnotatedTypeMirror castType = atypeFactory.getAnnotatedType(node);
+        AnnotationMirror castAnno = castType.getAnnotationInHierarchy(atypeFactory.UNKNOWNVAL);
+        AnnotationMirror exprAnno =
+                atypeFactory
+                        .getAnnotatedType(node.getExpression())
+                        .getAnnotationInHierarchy(atypeFactory.UNKNOWNVAL);
+
+        // Don't check a type cast if it is to a declared IntRange that covers the whole
+        // of the new type. I.e. do not warn if an @IntRange(from = Integer.MIN_VALUE,
+        // to = Integer.MAX_VALUE) int is casted to a @IntRange(from = Byte.MIN_VALUE,
+        // to = Byte.MAX_VALUE byte).
+        if (castAnno != null
+                && exprAnno != null
+                && atypeFactory.isIntRange(castAnno)
+                && atypeFactory.isIntRange(exprAnno)) {
+            Range castRange = ValueAnnotatedTypeFactory.getRange(castAnno);
+            Range exprRange = ValueAnnotatedTypeFactory.getRange(exprAnno);
+            if (exprRange.contains(castRange)) {
```

Comment:
I don't understand why this is needed.  If `castType` is a byte and is "byte everything", then isn't it a safe typecast?